pkg/operator/cockroachdb/controller.go

```diff
@@ -0,0 +1,573 @@
+	namespace string
+	spec      cockroachdbv1alpha1.ClusterSpec
+	ownerRef  metav1.OwnerReference
+}
+
+func newCluster(c *cockroachdbv1alpha1.Cluster, context *clusterd.Context) *cluster {
+	return &cluster{
+		context:   context,
+		namespace: c.Namespace,
+		spec:      c.Spec,
+		ownerRef:  clusterOwnerRef(c.Namespace, string(c.UID)),
+	}
+}
+
+func clusterOwnerRef(namespace, clusterID string) metav1.OwnerReference {
+	blockOwner := true
+	return metav1.OwnerReference{
+		APIVersion:         ClusterResource.Version,
+		Kind:               ClusterResource.Kind,
+		Name:               namespace,
+		UID:                types.UID(clusterID),
+		BlockOwnerDeletion: &blockOwner,
+	}
+}
+
+func (c *ClusterController) StartWatch(namespace string, stopCh chan struct{}) error {
+
+	resourceHandlerFuncs := cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	}
+
+	logger.Infof("start watching cockroachdb clusters in all namespaces")
+	watcher := opkit.NewWatcher(ClusterResource, namespace, resourceHandlerFuncs, c.context.RookClientset.CockroachdbV1alpha1().RESTClient())
+	go watcher.Watch(&cockroachdbv1alpha1.Cluster{}, stopCh)
+
+	return nil
+}
+
+func (c *ClusterController) onAdd(obj interface{}) {
+	clusterObj := obj.(*cockroachdbv1alpha1.Cluster)
+	logger.Infof("new cluster %s added to namespace %s", clusterObj.Name, clusterObj.Namespace)
+
+	cluster := newCluster(clusterObj, c.context)
+
+	if err := validateClusterSpec(cluster.spec); err != nil {
+		logger.Errorf("invalid cluster spec: %+v", err)
+		return
+	}
+
+	if err := c.createClientService(cluster); err != nil {
+		logger.Errorf("failed to create client service: %+v", err)
+		return
+	}
+
+	if err := c.createReplicaService(cluster); err != nil {
+		logger.Errorf("failed to create replica service: %+v", err)
+		return
+	}
+
+	if err := c.createPodDisruptionBudget(cluster); err != nil {
+		logger.Errorf("failed to create pod disruption budget: %+v", err)
+		return
+	}
+
+	if err := c.createStatefulSet(cluster); err != nil {
+		logger.Errorf("failed to create stateful set: %+v", err)
+		return
+	}
+
+	// retry to init the cluster until it succeeds or times out
+	err := wait.Poll(c.createInitRetryInterval, createInitTimeout, func() (bool, error) {
+		if err := c.initCluster(cluster); err != nil {
+			return false, nil
+		}
+
+		return true, nil
+	})
+	if err != nil {
+		logger.Errorf("failed to initialize cluster in namespace %s: %+v", cluster.namespace, err)
+		return
+	}
+
+	logger.Infof("succeeded creating and initializing cluster in namespace %s", cluster.namespace)
+}
+
+func (c *ClusterController) onUpdate(oldObj, newObj interface{}) {
+	_ = oldObj.(*cockroachdbv1alpha1.Cluster)
+	newCluster := newObj.(*cockroachdbv1alpha1.Cluster)
+	logger.Infof("cluster %s updated in namespace %s", newCluster.Name, newCluster.Namespace)
+}
+
+func (c *ClusterController) onDelete(obj interface{}) {
+	cluster := obj.(*cockroachdbv1alpha1.Cluster)
```

Comment:
Are you not supporting  delete operations in this patch?

Reply:
The teardown steps, owner refs and garbage collection clean up everything so the operator doesn't need to take specific steps on Deletion yet.  Let me know if I'm missing something here though.